failure to normalize associated types in supertraits

This example does not compile, but I believe it ought to ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=93dea30abcab62e4116dbec77ff37b81)):

```rust
use std::vec::Vec;

trait Foo<T> { }

trait Bar<T: Iterator>: Foo<T::Item> { }

fn a(x: &(impl Foo<u32> + ?Sized)) {
    
}

fn b(y: &dyn Bar<std::vec::IntoIter<u32>>) {
    a(y)
}

fn main() {
    
}
```

we get the error:

```
error[E0277]: the trait bound `dyn Bar<std::vec::IntoIter<u32>>: Foo<u32>` is not satisfied
  --> src/main.rs:12:5
   |
12 |     a(y)
   |     ^ the trait `Foo<u32>` is not implemented for `dyn Bar<std::vec::IntoIter<u32>>`
   |
note: required by `a`
  --> src/main.rs:7:1
   |
7  | fn a(x: &(impl Foo<u32> + ?Sized)) {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Note that changing `b` like so compiles ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7a50b8a49862872b76c18ddd41e89587)):

```rust
fn b(y: &(impl Bar<std::vec::IntoIter<u32>> + ?Sized)) { }
```

--- a/tests/ui/associated-types/normalize-supertrait-projection-in-dyn-61083.rs
+++ b/tests/ui/associated-types/normalize-supertrait-projection-in-dyn-61083.rs
@@ -1,0 +1,28 @@
+//! Regression test for <https://github.com/rust-lang/rust/issues/61083>.
+//!
+//! An associated type projection in a supertrait bound (`Bar<T>: Foo<T::Item>`)
+//! failed to normalize when the `Bar` bound was reached through a trait object,
+//! so passing the object to a function expecting `Foo<u32>` was rejected.
+
+//@ revisions: current next
+//@ ignore-compare-mode-next-solver (explicit revisions)
+//@[next] compile-flags: -Znext-solver
+//@ check-pass
+
+trait Foo<T> {}
+
+trait Bar<T: Iterator>: Foo<T::Item> {}
+
+fn a(_x: &(impl Foo<u32> + ?Sized)) {}
+
+// The `dyn` form is the one that used to fail to normalize `T::Item` to `u32`.
+fn b(y: &dyn Bar<std::vec::IntoIter<u32>>) {
+    a(y)
+}
+
+// The equivalent `impl Trait` form always compiled; keep it so both paths stay pinned.
+fn c(y: &(impl Bar<std::vec::IntoIter<u32>> + ?Sized)) {
+    a(y)
+}
+
+fn main() {}

